### ISSUE TITLE
py_wheel and py_package of internal library results in package of pip transitive targets

### ISSUE THREAD

**Issue by fferroni** [bazelbuild/rules_python#1300]
# 🐞 bug report

### Affected Rule

`py_wheel` generates wheels that contain transitive pip package dependencies inside the wheel. This yields extremely large wheels.

### Is this a regression?

Not sure, but it's not how a Pythonic wheel packages things.

### Description

I have a py_library that depends transitively on some external pip packages, defined in the various libraries as `requirement("...")` in the `deps` list.
When I use `py_package` on my repository library, and then feed the `py_package` target to `py_wheel`, it generates a `whl` that also contains all the pip-fetched repositories. I have some dependencies like torch which have huge binaries, leading my (pure python) library to be enormous.

## 🔬 Minimal Reproduction

Using rules_python `0.23.1`, Bazel `5.4.0`, and standard provided toolchain and interpreter (`3.8.13`).

`test.py`:
```
import torch

def main():
    print("hello world")

if __name__ == "__main__":
    main()
```

`BUILD`:
```
py_binary(
    name = "test",
    srcs = ["test.py"],
    deps = [requirement("torch")],
)

py_package(
    name = "test_pkg",
    packages = [],
    deps = [":test"],
)

py_wheel(
    name = "test_whl",
    distribution = "test_pip",
    python_requires = ">=3.8",
    python_tag = "py3",
    requires = [
        "torch",
    ],
    version = "0.0.1",
    deps = [
        ":test_pkg",
    ],
)
```

The generated wheel is huge, and if you change the extension from `.whl` to `.zip` and check the internal files, there's lots of `torch` specific code (and any other deps from `requirement("...")` targets).
Is this desired behaviour?

**Comment by rickeylev** [comment]
I believe the problem is `packages = []`. An empty list is treated the same as not specifying it -- all transitive deps are collected. You need to specify a non-empty list to restrict what is included.

**Comment by pstradomski** [comment]
> I believe the problem is `packages = []`. An empty list is treated the same as not specifying it -- all transitive deps are collected. You need to specify a non-empty list to restrict what is included.

Given requirements now IIUC add extra providers, maybe we could strip them automatically from the depsets; though I'm not sure about it - and of course it's possible somebody actually wants to package some deps.

For better user experience, maybe we should disallow empty `packages`?

**Comment by fferroni** [comment]
> I believe the problem is `packages = []`. An empty list is treated the same as not specifying it -- all transitive deps are collected. You need to specify a non-empty list to restrict what is included.

OK - in that case, what would be the correct list of packages to provide? I simply want to include everything that is in the my repository (in this minimal example it's ":test", but I'd like the transitive dependencies that are local to the repository).

**Comment by chrislovecnm** [comment]
@rickeylev ping

**Comment by fferroni** [comment]
Hi, wondering if there is a guide on how to properly do this then? @pstradomski @rickeylev

**Comment by rickeylev** [comment]
> I simply want to include everything that is in the my repository

Looking at the implementation, this is sort of possible today. It looks like the `packages` arg is a prefix filter. 

The underlying code is something like:

```
<for each entry in packages>
package = p.replace(".", "/")
wheel_path = <File.short_path, with "../<repo>/" removed>
if wheel_path.startswith(package): <include in output>
```

So, if all your code is under a particular directory, then you can specify those, and anything under them will be included. Code from external repos might be included, too, though, depending on if your repo shares a common prefix with them. Not really ideal :\.

The docs sort of say this, "Sub-packages are automatically included", but I think if they made it more clear it's a prefix-match, that would help explain behavior.

We should have a way to distinguish between own repo vs external dependencies, though.

Maybe allow "@<name>//" label notation? This would match other contexts in Bazel. So if you wanted to restrict to files just in your own package, you'd do: `py_package(packages=["@//:__subpackages__"])`

Or some such. The key part being, the rule receives a Label object, which it can then compare to `File.owner` when considering what to include.

-----
  

Note that the inputs to py_package.deps are just files. If you want to work around this today, then you can use custom starlark to do so. Something like:

```
def impl(ctx):
  files = []
  for f in ctx.files.deps:
    if f.owner.workspace_name == <whatever>:
      files.append(f)
  return [DefaultInfo(files=depset(files))
```

I think `rules_pkg` can also do various transforms.
